### ISSUE TITLE
[no-misused-promises] reports for .forEach but not for .map

### ISSUE THREAD

**Issue by avalanche1** [typescript-eslint/typescript-eslint#3116]
This is hit with an error:
```ts
[1, 2, 3].forEach(async value => {
  await doSomething(value);
});
```
`Promise returned in function argument where a void return was expected`

Whereas this doesn't flag:
```ts
[1, 2, 3].map(async value => {
  await doSomething(value);
});
```

AFAIK, .forEach and .map treat promises the same way.

**Comment by avalanche1** [comment]
And a side question:
```ts
for (const value of [1, 2, 3]) {
  await doSomething(value);
}
```
Why is this considiered correct? In my understanding, the side effects will be the same as with .forEach: promises will be fulfilled in random order.

**Comment by bradzacher** [comment]
this is correct.
`.foreach` is not an async method, and does not return a value.
that means passing an async function into it makes it impossible to `await` the result.
i.e. `.foreach` *always* returns `void`

OTOH, `.map` returns the promises.
i.e. `.map` returns `Array<Promise<...>>`
So you can await the results appropriately.

----

> Why is this considiered correct? In my understanding, the side effects will be the same as with .forEach: promises will be fulfilled in random order.

Incorrect.
The `await` in the loop means that the loop will not iterate until the promise returned by `doSomething` resolves.
This is different to `.forEach`, because `.forEach` applies no awaits at all.

I.e. this code is the same as `.forEach`:
```ts
[1,2,3].forEach(value => doSomething(value));

// same as

for (const value of [1, 2, 3]) {
  doSomething(value);
}


const result = [1,2,3].map(async value => doSomething(value));
// wait for the async workload to finish
await Promise.all(result);

// same as

const result = [];
for (const value of [1,2,3]) {
  result.push(doSomething(value));
}
// wait for the async workload to finish
await Promise.all(result);
```

**Comment by avalanche1** [comment]
If I want to use await in .forEach - what should I do?
eg
```ts
[1,2,3].forEach(await x => {
  let y = await get_value(x);
  send_somewhere(y);
});
```

**Comment by bradzacher** [comment]
You should avoid doing asynchronous work serially, as that kind of defeats the purpose of async.

Bit if you really want to intentionally do this pattern, then use a for of and an await.

**Comment by avalanche1** [comment]
I don't want to  do asynchronous work serially.
My intention is: *To await some value and use it in later processing. Do this for a sequence of basis values.*
Just like my example above.
Each iteration is completely independent from another. *I don't need one iteration to finish to start the next.*

Thus, I consider .forEach to be a perfect tool for my use case.

Thus, I consider `no-misused-promises` triggering in this scenario to be a mistake.

**Comment by avalanche1** [comment]
Currently, I have to suppress it with `// eslint-disable-next-line` and write an `explain` every time I do it.
I'd like it to either not trigger; or have a separate turn-off-able option for this scenario.

**Comment by bradzacher** [comment]
> My intention is: To await some value and use it in later processing. Do this for a sequence of basis values.

Your intention is to fire off a set of async workloads and then do something when they complete. 

So I've already given you the correct code to use.
Use either `.map` or the equivalent for of. That way you can actually await the workload whilst they all proceed in parallel.



`.forEach` ***always*** returns `void`
Which means you can ***never*** await it.
Which means you cannot know when the async work is done.

**Comment by avalanche1** [comment]
> Your intention is to fire off a set of async workloads and then do something when they complete.

I do not want to do anything after they complete. They are the finish effect of some event. Nothing else is gonna happen until the next event.

That is why *I do not*  want to *await* for `.forEach` completion.

**Comment by avalanche1** [comment]
That is why returning in `.map` is relevant; but returning in `.forEach` is not.

So I would like this rule not to report `.forEach`.
